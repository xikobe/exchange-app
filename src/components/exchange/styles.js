import styled from '@emotion/styled';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  max-width: 500px;
  width: 100%;
  margin: 0 auto;
`;
