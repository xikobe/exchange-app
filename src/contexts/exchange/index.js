import React, {
  createContext, useContext, useState,
} from 'react';
import PropTypes from 'prop-types';
import { useRates } from '../../queries/rates';

const ExchangeContext = createContext();

export const useExchangeContext = () => useContext(ExchangeContext);

const ExchangeProvider = ({ children }) => {
  const [availableCurrencies, setAvailableCurrencies] = useState(['USD', 'EUR', 'GBP']);
  const [activeCurrency, setActiveCurrency] = useState('USD');
  const [tradeCurrency, setTradeCurrency] = useState('EUR');
  const { data: activeRate } = useRates({ base: activeCurrency });
  const { data: tradeRate } = useRates({ base: tradeCurrency });

  const getCurrency = (isTrade) => (isTrade ? tradeCurrency : activeCurrency);

  const handleSwitchCurrency = () => {
    const tempActive = activeCurrency;
    const tempTrade = tradeCurrency;
    setActiveCurrency(tempTrade);
    setTradeCurrency(tempActive);
  }

  const handleChangeCurrency = (isTrade, value) => {
    if(isTrade) {
      return setTradeCurrency(value);
    }

    return setActiveCurrency(value);
  }

  return (
    <ExchangeContext.Provider value={{
      availableCurrencies,
      activeCurrency,
      tradeCurrency,
      getCurrency,
      handleSwitchCurrency,
      activeRate: activeRate?.rates[tradeCurrency],
      tradeRate: tradeRate?.rates[activeCurrency],
      handleChangeCurrency,
    }}
    >
      { children}
    </ExchangeContext.Provider>
  );
};

ExchangeProvider.propTypes = {
  children: PropTypes.node,
};

export default ExchangeProvider;
