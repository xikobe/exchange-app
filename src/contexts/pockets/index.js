import React, {
  createContext, useContext, useState,
} from 'react';
import PropTypes from 'prop-types';

const PocketsContext = createContext();

export const usePocketsContext = () => useContext(PocketsContext);

const PocketsProvider = ({ children }) => {
  const [pocket, setPocket] = useState({
    USD: 100,
    EUR: 90,
    GBP: 70,
  });
  const [inputValue, setInputValue] = useState(0);
  const [tradeValue, setTradeValue] = useState(0);

  const getBalance = (currency) => pocket[currency];
  const getInputValue = (isTrade) => (isTrade ? tradeValue : inputValue);
  const makeExchange = ({ activeCurrency, tradeCurrency }) => {
    setPocket((prevPocket) => ({
      ...prevPocket,
      [activeCurrency]: prevPocket[activeCurrency] - inputValue,
      [tradeCurrency]: prevPocket[tradeCurrency] + tradeValue,
    }));
  };

  return (
    <PocketsContext.Provider value={{
      pocket,
      getBalance,
      inputValue,
      tradeValue,
      setInputValue,
      setTradeValue,
      getInputValue,
      makeExchange,
    }}
    >
      {children}
    </PocketsContext.Provider>
  );
};

PocketsProvider.propTypes = {
  children: PropTypes.node,
};

export default PocketsProvider;
