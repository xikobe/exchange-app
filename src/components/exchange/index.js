import React from 'react';
import { Wrapper } from './styles';
import RateCard from '../rate-card';
import RateValue from '../rate-value';
import ExchangeContext from '../../contexts/exchange';
import SwitchCurrency from '../switch-currency';
import SubmitExchange from './submit-exchange';

const Exchange = () => (
  <Wrapper>
    <ExchangeContext>
      <RateCard />
      <RateValue />
      <SwitchCurrency />
      <RateCard trade />
      <SubmitExchange />
    </ExchangeContext>
  </Wrapper>
);

export default Exchange;
