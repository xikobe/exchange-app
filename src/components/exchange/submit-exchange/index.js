import React from 'react';
import { Button, Text } from '@chakra-ui/core';
import { usePocketsContext } from '../../../contexts/pockets';
import { useExchangeContext } from '../../../contexts/exchange';
import { Wrapper } from './styles';

const SubmitExchange = () => {
  const { makeExchange } = usePocketsContext();
  const { activeCurrency, tradeCurrency, exchangeError } = useExchangeContext();

  const handleOnClick = () => {
    makeExchange({ activeCurrency, tradeCurrency });
  };

  return (
    <Wrapper>
      <Button variantColor="pink" marginBottom="20px" disabled={activeCurrency === tradeCurrency || !!exchangeError} onClick={handleOnClick}>Exchange</Button>
      { exchangeError && <Text bg="red.500" padding="10px" color="white">{exchangeError}</Text> }
    </Wrapper>
  );
};

export default SubmitExchange;
