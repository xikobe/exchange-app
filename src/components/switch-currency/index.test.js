import React from 'react';
import { shallow } from 'enzyme';
import SwitchCurrency from './index';
import { useExchangeContext } from '../../contexts/exchange';

jest.mock('../../contexts/exchange');

useExchangeContext.mockImplementation(() => ({}));

describe('SwitchCurrency', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<SwitchCurrency />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Button').prop('isLoading')).toBeFalsy();
  });

  it('renders loading state correctly', () => {
    useExchangeContext.mockImplementation(() => ({
      isLoadingRate: true,
    }));

    const wrapper = shallow(<SwitchCurrency />);

    expect(wrapper.find('Button').prop('isLoading')).toBe(true);
  });

  it('triggers the handleSwitchCurrency method on button click', () => {
    const mockHandleSwitchCurrency = jest.fn();
    useExchangeContext.mockImplementation(() => ({
      handleSwitchCurrency: mockHandleSwitchCurrency,
    }));

    const wrapper = shallow(<SwitchCurrency />);
    wrapper.find('Button').simulate('click');
    expect(mockHandleSwitchCurrency).toHaveBeenCalled();
  });
});
