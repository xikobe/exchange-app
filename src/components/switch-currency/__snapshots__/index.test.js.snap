// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchCurrency renders correctly 1`] = `
<Styled(div)>
  <Button
    icon="repeat"
    loadingText="Fetching rate"
    variant="outline"
    variantColor="blue"
  >
    <RateValue />
    <Icon
      color="teal"
      name="repeat"
    />
  </Button>
</Styled(div)>
`;
